build.gradle

```diff
@@ -137,6 +137,7 @@ configure(opentelemetryProjects) {
                 jmh_core                : "org.openjdk.jmh:jmh-core:${jmhVersion}",
                 jmh_bytecode            : "org.openjdk.jmh:jmh-generator-bytecode:${jmhVersion}",
                 jsr305                  : "com.google.code.findbugs:jsr305:${findBugsJsr305Version}",
+                oshi                    : "com.github.oshi:oshi-core:5.1.1",
```

Comment:
JNA libraries tend to be pretty bloated since they basically duplicate a native interface into Java, and for cross-platform ones have a ton of these. This seems to be 600K along with JNA itself which is 3MB+ (SDK is just 250K for reference). JDK's monitoring beans provide a lot of useful information so maybe just focus on them for now and can consider adding the huge dependency in a followup if the delta in terms of metrics seems useful?

Reply:
I guess if it's just about system metrics this could be helpful. But it seems more useful to users to focus on JVM metrics first? None of the popular Java metrics libraries seem to use oshi but provide very useful runtime metrics without much binary cost - such metrics should basically "always be enabled" on an instrumented app while this seems much more specialized and should just add the metrics that can't be gotten for free from the JDK.

https://github.com/dropwizard/metrics/tree/release/4.1.x/metrics-jvm/src/main/java/com/codahale/metrics/jvm
https://github.com/prometheus/client_java/tree/master/simpleclient_hotspot/src/main/java/io/prometheus/client/hotspot
https://github.com/micrometer-metrics/micrometer/tree/b486e1261cade606eb0ff826fbae318f6d0c9e86/micrometer-core/src/main/java/io/micrometer/core/instrument/binder/system

Reply:
>  this seems much more specialized and should just add the metrics that can't be gotten for free from the JDK.

That's correct, and this is the precisely the idea: to provide a set of complete system-wide metrics (and was the reason why I wanted to initially put it here, as a separate, extra artifact that some users can use). 

@trask is there a plan to support external instrumentation jars? In that case, this could be provided at runtime. Also, what's your take on having Oshi as a dependency?

Reply:
Many of the metrics here, like system cpu load, are available from the JDK though. And we definitely wouldn't want double instrumentation of the same metric if people added both, e.g., `instrumentation-jvm` and `instrumentation-system`. So recommendation was to start with the less complex, free stuff since that gives a better picture of what's needed in the complex, dependency-heavy version. So this feels like it's the wrong order to be addressing the issue in IMO. Does that make sense?

Reply:
This will be a rather small codebase, so I don't see the need to postpone the discussion on what to use/report ;) I'd argue that it's not a wrong order situation, as I mentioned before, the idea is precisely to provide system-wide metrics.

> we definitely wouldn't want double instrumentation of the same metric if people added both

It's an interesting scenario to have them separated. But in that case I'd rather have the system-wide disabled or something similar. Unless there's a clear path that justifies having them separated (in which case I'd be fine with the separation).

Reply:
> @trask is there a plan to support external instrumentation jars?

ya, we need to support users writing/bringing their own instrumentation somehow, but we haven't gotten to that yet

> Also, what's your take on having Oshi as a dependency?

well, the auto agent jar is already 28mb, so what's another 3mb? 😭

(side note: maybe we should use the `all` classifier after all, to be clear that it's the kitchen sink of distributions)

I agree that it would be nice to split this into two modules, one that only has only JVM (mbean) metrics, and one that has only system (oshi) metrics.

The JVM metric module could detect if the system metrics module is on the classpath and disable anything that's duplicative with the system metrics (e.g. don't capture cpu `total` via JVM metrics, since system metrics captures cpu `system` + `user`).

Reply:
Having the jvm module disable metrics if oshi is being used is a nice idea.

Size-wise, let's not forget we are trying to move towards also supporting manual instrumentation, not just agent. Though 10% of the entire agent seems pretty hefty FWIW. System metrics also have the issue that multiple jvms on the same system will also provide redundant metrics, common with containers. That should be taken into account too when evaluating the cost, it does take actual RAM to load large ffi modules. I would expect of packaged into the agent jar it's not enabled by default because of this potentially exploding cost to users.

That's enough high level context, for this PR how about removing the GC metrics and only take stuff from oshi? Can add GC metrics to a separate jvm metrics module

Reply:
Ah found that we actually already have such a class too

https://github.com/open-telemetry/opentelemetry-java/blob/596617dcbe7763435573a3d02eba4a5986f7dcd8/contrib/runtime_metrics/src/main/java/io/opentelemetry/contrib/metrics/runtime/GarbageCollector.java#L66

Reply:
> I agree that it would be nice to split this into two modules, one that only has only JVM (mbean) metrics, and one that has only system (oshi) metrics.

That's fine by me, as long as one module is for JVM-specific metrics (the current `metrics` module in contrib, with a few metric additions) and have this one is for System-wide metrics (which could be disabled by default). However, skipping the Oshi dependency wouldn't be avoided ;) 

Btw, related effort: https://github.com/open-telemetry/opentelemetry-specification/issues/651